File: 2 - Stack/Php/stack.php
<?php

class Stack {
    private $data;
    private $capacity;

    public function __construct($capacity) {
        $this->capacity = $capacity;
        $this->data = [];
    }

    public function isFull() {
        return count($this->data) == $this->capacity;
    }

    public function isEmpty() {
        return count($this->data) == 0;
    }

    public function push($value) {
        if ($this->isFull()) {
            $this->resize();
        }
        array_push($this->data, $value);
        echo "Pushed $value onto stack.\n";
    }

    public function pop() {
        if ($this->isEmpty()) {
            echo "Stack is EMPTY! Cannot pop.\n";
            return;
        }
        $poppedValue = array_pop($this->data);
        echo "Popped $poppedValue from stack.\n";
    }

    public function topValue() {
        if ($this->isEmpty()) {
            echo "Stack is EMPTY! No top value.\n";
            return null;
        }
        return end($this->data);
    }

    public function peek($n) {
        if ($this->isEmpty()) {
            echo "Stack is EMPTY! No element to peek.\n";
            return null;
        }
        $index = count($this->data) - $n;
        if ($index < 0 || $index >= count($this->data)) {
            echo "Invalid peek($n). Not enough elements.\n";
            return null;
        }
        return $this->data[$index];
    }

    public function display() {
        if ($this->isEmpty()) {
            echo "Stack is EMPTY!\n";
            return;
        }
        echo "Stack contents (top -> bottom):\n";
        for ($i = count($this->data) - 1; $i >= 0; $i--) {
            echo "[" . ($i + 1) . "]: " . $this->data[$i] . "\n";
        }
        echo "\n";
    }

    private function resize() {
        $this->capacity *= 2;
        echo "Stack resized.\n";
    }
}

$books = new Stack(5);

$books->push("Mathematics");
$books->push("Calculus");
$books->push("Physics");
$books->push("Web Design");
$books->push("Data Structure");

$books->display();

echo "Top of stack: " . $books->topValue() . "\n";
echo "Peek(2): " . $books->peek(2) . "\n";

$books->pop();

$books->display();
?>

File: 2 - Stack/Java/stack/stack.java
import java.util.ArrayList;

public class stack {
    private ArrayList<String> data;
    private int capacity;

    public stack(int capacity) {
        this.capacity = capacity;
        this.data = new ArrayList<>(capacity);
    }

    public boolean isFull() {
        return data.size() == capacity;
    }

    public boolean isEmpty() {
        return data.isEmpty();
    }

    public void push(String value) {
        if (isFull()) {
            resize();
        }
        data.add(value);
        System.out.println("Pushed " + value + " onto stack.");
    }

    public void pop() {
        if (isEmpty()) {
            System.out.println("stack is EMPTY! Cannot pop.");
            return;
        }
        String poppedValue = data.remove(data.size() - 1);
        System.out.println("Popped " + poppedValue + " from stack.");
    }

    public String topValue() {
        if (isEmpty()) {
            System.out.println("stack is EMPTY! No top value.");
            return null;
        }
        return data.get(data.size() - 1);
    }

    public String peek(int n) {
        if (isEmpty()) {
            System.out.println("stack is EMPTY! No element to peek.");
            return null;
        }
        int index = data.size() - n;
        if (index < 0 || index >= data.size()) {
            System.out.println("Invalid peek(" + n + "). Not enough elements.");
            return null;
        }
        return data.get(index);
    }

    public void display() {
        if (isEmpty()) {
            System.out.println("stack is EMPTY!");
            return;
        }
        System.out.println("stack contents (top -> bottom):");
        for (int i = data.size() - 1; i >= 0; i--) {
            System.out.println("[" + (i + 1) + "]: " + data.get(i));
        }
        System.out.println();
    }

    private void resize() {
        capacity *= 2;
        System.out.println("stack resized.");
    }

    public static void main(String[] args) {
        stack books = new stack(5);

        books.push("Mathematics");
        books.push("Calculus");
        books.push("Physics");
        books.push("Web Design");
        books.push("Data Structure");

        books.display();

        System.out.println("Top of stack: " + books.topValue());
        System.out.println("Peek(2): " + books.peek(2));

        books.pop();

        books.display();
    }
}